Update: NPE

Got a NPE while concurrently update and insert on same page

`java.util.concurrent.ExecutionException: java.lang.RuntimeException: java.lang.NullPointerException
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at herddb.server.DirectMultipleConcurrentUpdatesTest.performTest(DirectMultipleConcurrentUpdatesTest.java:201)
	at herddb.server.DirectMultipleConcurrentUpdatesTest.testWithCheckpoints(DirectMultipleConcurrentUpdatesTest.java:80)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:379)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:340)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:413)
Caused by: java.lang.RuntimeException: java.lang.NullPointerException
	at herddb.server.DirectMultipleConcurrentUpdatesTest$1.run(DirectMultipleConcurrentUpdatesTest.java:194)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at herddb.core.PageSet.setPageDirty(PageSet.java:107)
	at herddb.core.PageSet.setPageDirty(PageSet.java:117)
	at herddb.core.TableManager.applyUpdate(TableManager.java:1331)
	at herddb.core.TableManager.apply(TableManager.java:1180)
	at herddb.core.TableManager$6.accept(TableManager.java:963)
	at herddb.core.TableManager.lambda$accessTableData$42(TableManager.java:2475)
	at herddb.core.TableManager$$Lambda$73/1306501541.execute(Unknown Source)
	at herddb.utils.BatchOrderedExecutor.finish(BatchOrderedExecutor.java:72)
	at herddb.core.TableManager.accessTableData(TableManager.java:2501)
	at herddb.core.TableManager.executeUpdate(TableManager.java:939)
	at herddb.core.TableManager.executeStatement(TableManager.java:499)
	at herddb.core.TableSpaceManager.executeTableAwareStatement(TableSpaceManager.java:1053)
	at herddb.core.TableSpaceManager.executeStatement(TableSpaceManager.java:988)
	at herddb.model.planner.SimpleUpdateOp.execute(SimpleUpdateOp.java:52)
	at herddb.core.TableSpaceManager.executePlannedOperationStatement(TableSpaceManager.java:1033)
	at herddb.core.TableSpaceManager.executeStatement(TableSpaceManager.java:991)
	at herddb.core.DBManager.executeStatement(DBManager.java:596)
	at herddb.core.DBManager.executePlan(DBManager.java:667)
	at herddb.core.TestUtils.executeUpdate(TestUtils.java:49)
	at herddb.server.DirectMultipleConcurrentUpdatesTest$1.run(DirectMultipleConcurrentUpdatesTest.java:163)
	... 5 more`

--- a/herddb-core/src/main/java/herddb/core/PageSet.java
+++ b/herddb-core/src/main/java/herddb/core/PageSet.java
@@ -24,7 +24,9 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.LongAdder;
+import java.util.logging.Level;
 import java.util.logging.Logger;
 
 import herddb.model.Record;
@@ -42,7 +44,7 @@ public final class PageSet {
     private static final Logger LOGGER = Logger.getLogger(PageSet.class.getName());
 
     /** Active stored pages (active/size,average-record-size,dirt,hasDeletions)*/
-    private final Map<Long,DataPageMetaData> activePages = new ConcurrentHashMap<>();
+    private final ConcurrentMap<Long,DataPageMetaData> activePages = new ConcurrentHashMap<>();
 
     public static final class DataPageMetaData {
 
@@ -129,9 +131,16 @@ public final class PageSet {
         return activePages.values().stream().mapToInt(meta -> meta.dirt.sum() > 0 ? 1 : 0).sum();
     }
 
-    void pageCreated(Long pageId, DataPage page) {
-        final DataPageMetaData old = activePages.put(pageId, new DataPageMetaData(page));
+    /**
+     * @throws IllegalStateException if a page with give id already existed
+     */
+    void pageCreated(Long pageId, DataPage page) throws IllegalStateException {
+        /* Don't really add page if already exists */
+        final DataPageMetaData old = activePages.putIfAbsent(pageId, new DataPageMetaData(page));
         if (old != null) {
+            LOGGER.log(Level.SEVERE,
+                    "Detected concurrent creation of page " + page.pageId + ", writable: " + page.writable);
+
             throw new IllegalStateException("Creating a new page already existing! Page " + pageId);
         }
     }

--- a/herddb-core/src/main/java/herddb/core/TableManager.java
+++ b/herddb-core/src/main/java/herddb/core/TableManager.java
@@ -757,6 +757,10 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
              * concurrent thread or because it wasn't present in the first place.
              */
 
+
+            /* Set the new page as a fully active page */
+            pageSet.pageCreated(page.pageId, page);
+
             /* Remove it from "new" pages */
             DataPage remove = newPages.remove(page.pageId);
             if (remove == null) {
@@ -764,9 +768,6 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
                         "Detected concurrent flush of page " + page.pageId + ", writable: " + page.writable);
                 throw new IllegalStateException("page " + page.pageId + " is not a new page!");
             }
-
-            /* Set the new page as a fully active page */
-            pageSet.pageCreated(page.pageId, page);
 
             page.writable = false;
 
@@ -940,7 +941,7 @@ public final class TableManager implements AbstractTableManager, Page.Owner {
             @Override
             public void accept(Record actual) throws StatementExecutionException, LogNotAvailableException, DataStorageManagerException {
                 byte[] newValue = function.computeNewValue(actual, context, tableContext);
-                
+
                 if (indexes != null) {
                     try {
                         DataAccessor values = new Record(actual.key, Bytes.from_array(newValue)).getDataAccessor(table);

